synapse/storage/databases/main/end_to_end_keys.py

```diff
@@ -701,15 +752,35 @@ def _claim_e2e_one_time_keys(txn):
                 " WHERE user_id = ? AND device_id = ? AND algorithm = ?"
                 " LIMIT 1"
             )
+            fallback_sql = (
+                "SELECT key_id, key_json, used FROM e2e_fallback_keys_json"
+                " WHERE user_id = ? AND device_id = ? AND algorithm = ?"
+                " LIMIT 1"
+            )
             result = {}
             delete = []
+            used_fallbacks = []
             for user_id, device_id, algorithm in query_list:
                 user_result = result.setdefault(user_id, {})
                 device_result = user_result.setdefault(device_id, {})
                 txn.execute(sql, (user_id, device_id, algorithm))
+                found = False
                 for key_id, key_json in txn:
+                    found = True
                     device_result[algorithm + ":" + key_id] = key_json
                     delete.append((user_id, device_id, algorithm, key_id))
+                if not found:
```

Comment:
Isn't this only going to iterate once anyway since there's the `LIMIT 1`? I feel like the `found` logic is confusing and should be able to be simplified, but maybe not.

Reply:
Yup.  This is because `txn` seems to be an iterable, but I can't figure out if there's an easy way of just fetching the first (and only) result other than either a for loop or a `[r for r in txn]`.  Maybe converting to a list would be clearer?  Or is there a better way?

Reply:
We seem to use `row = txn.fetchone()` in many places (note that `row` will be `None` if the value isn't found).